lib/improver/spotdata_new/spot_extraction.py

```diff
@@ -0,0 +1,247 @@
+
+    @staticmethod
+    def build_diagnostic_cube(neighbour_cube, diagnostic_cube,
+                              spot_values):
+        """
+        Builds a spot data cube containing the extracted diagnostic values.
+
+        Args:
+            neighbour_cube (iris.cube.Cube):
+                This cube is needed as a source for information about the spot
+                sites which needs to be included in the spot diagnostic cube.
+            diagnostic_cube (iris.cube.Cube):
+                The cube is needed to provide the name and units of the
+                diagnostic that is being processed.
+            spot_values (np.array):
+                An array containing the diagnostic values extracted for the
+                required spot sites.
+        Returns:
+            neighbour_cube (iris.cube.Cube):
+                A spot data cube containing the extracted diagnostic data.
+        """
+
+        neighbour_cube = build_spotdata_cube(
+            spot_values, diagnostic_cube.name(), diagnostic_cube.units,
+            neighbour_cube.coord('altitude').points,
+            neighbour_cube.coord(axis='y').points,
+            neighbour_cube.coord(axis='x').points,
+            neighbour_cube.coord('wmo_id').points)
+        return neighbour_cube
+
+    def process(self, neighbour_cube, diagnostic_cube):
+        """
+        Create a spot data cube containing diagnostic data extracted at the
+        coordinates provided by the neighbour cube.
+
+        .. See the documentation for more details about the inputs and output.
+        .. include:: extended_documentation/spot_extraction_examples.rst
+
+        Args:
+            neighbour_cube (iris.cube.Cube):
+                A cube containing information about the spot data sites and
+                their grid point neighbours.
+            diagnostic_cube (iris.cube.Cube):
+                A cube of diagnostic data from which spot data is being taken.
+        Returns:
+            spotdata_cube (iris.cube.Cube):
+                A cube containing diagnostic data for each spot site, as well
+                as information about the sites themselves.
+        """
+
+        # Check we are using a matched neighbour/diagnostic cube pair
+        if not check_cube_compatibility(self.grid_metadata_identifier,
+                                        neighbour_cube, diagnostic_cube):
+            msg = ("The grid_metadata_identifier attributes do not match "
+                   "between the neighbour cube and the diagnostic cube. "
+                   "Cubes may not be for the same grid/model.")
+            raise ValueError(msg)
+
+        coordinate_cube = self.extract_coordinates(neighbour_cube)
+
+        # Deal with leading dimensions such as thresholds, realizations, etc.
+        data_cubes = iris.cube.CubeList()
+        for cube in diagnostic_cube.slices(
+                [diagnostic_cube.coord(axis='x').name(),
+                 diagnostic_cube.coord(axis='y').name()]):
```

Comment:
Again, why are we going against the convention and using (x, y) slice ordering here?

Reply:
Because the indices from neighbour finding are in this order, and if we are using indices they have to match. As mentioned earlier, they are in x-y order in neighbour finding to align with some of the cartopy functions that are used there.

Reply:
OK.